File: apps/api/src/lib/email.ts
import { env } from '../config/env.js';

export interface EmailProvider {
  sendInvitation(to: string, name: string, inviteLink: string): Promise<void>;
}

class ConsoleEmailProvider implements EmailProvider {
  async sendInvitation(to: string, name: string, inviteLink: string): Promise<void> {
    console.log('\n========== EMAIL INVITATION ==========');
    console.log(`To: ${to}`);
    console.log(`Name: ${name}`);
    console.log(`Link: ${inviteLink}`);
    console.log('=======================================\n');
  }
}

class ResendEmailProvider implements EmailProvider {
  async sendInvitation(to: string, name: string, inviteLink: string): Promise<void> {
    const apiKey = process.env.RESEND_API_KEY;
    if (!apiKey) throw new Error('RESEND_API_KEY not configured');

    await fetch('https://api.resend.com/emails', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        from: env.EMAIL_FROM,
        to,
        subject: 'Undangan Bergabung - AISI',
        html: `
          <h2>Assalamu'alaikum ${name},</h2>
          <p>Anda diundang untuk bergabung ke Aplikasi AISI.</p>
          <p><a href="${inviteLink}">Klik di sini untuk membuat password</a></p>
          <p>Link berlaku 7 hari.</p>
        `,
      }),
    });
  }
}

export const emailProvider: EmailProvider = process.env.RESEND_API_KEY
  ? new ResendEmailProvider()
  : new ConsoleEmailProvider();

File: apps/api-go/internal/email/email.go
package email

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/dakwah-depok/aisi/apps/api-go/internal/config"
)
type Sender struct { from, apiKey string }
func New(c config.Config) Sender { return Sender{from: c.EmailFrom, apiKey: c.ResendAPIKey} }
func (s Sender) SendInvitation(ctx context.Context, to, name, url string) error {
	if s.apiKey == "" { fmt.Printf("[email console] to=%s name=%s invitation=%s\n", to, name, url); return nil }
	body, _ := json.Marshal(map[string]string{"from": s.from, "to": to, "subject": "Undangan AISI", "html": "<p>Assalamu'alaikum "+name+",</p><p><a href=\""+url+"\">Atur password akun</a></p>"})
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.resend.com/emails", bytes.NewReader(body)); if err != nil { return err }
	req.Header.Set("Authorization", "Bearer "+s.apiKey); req.Header.Set("Content-Type", "application/json")
	res, err := http.DefaultClient.Do(req); if err != nil { return err }; defer res.Body.Close()
	if res.StatusCode >= 300 { return fmt.Errorf("Resend mengembalikan %s", res.Status) }; return nil
}
